tensorflow/core/common_runtime/zen_layout_pass.cc

```diff
@@ -0,0 +1,1196 @@
+      if (ZenOpNodeRewrite(g, n, rewrite_record, n_reorder) == OkStatus()) {
+        VLOG(1) << "ZenLayoutRewritePass::ZenOpUpdate: Node " << op_name
+                << " rewritten with ZenOp " << rewrite_record->zen_op_name;
+        result = true;
+      } else {
+        // Rewriting the node with Zen op failed. Hence the existing node will
+        // be there with graph for inference.
+        VLOG(1) << "ZenLayoutRewritePass::ZenOpUpdate: Failed to rewrite node "
+                << node_name << " with Zen op " << op_name;
+      }
+    }
+  }
+  return result;
+}
+
+// Method to find whether the graph has inference ops only. It returns error
+// status if the graph has training ops.
+Status ZenLayoutRewritePass::AreAllInferenceOps(std::unique_ptr<Graph> *g) {
+  std::vector<Node *> order;
+  GetReversePostOrder(**g, &order);
+  for (Node *n : order) {
+    if (!n->IsOp()) {
+      continue;
+    }
+    for (auto op = tf_training_ops_.cbegin(); op != tf_training_ops_.cend();
+         ++op) {
+      if (n->type_string().find(*op) != string::npos) {
+        return Status(error::Code::UNIMPLEMENTED,
+                      "Training operation found! Currently TF-ZenDNN "
+                      "does not support training. Set environment "
+                      "variable TF_ENABLE_ZENDNN_OPTS to '0' for training.");
+      }
+    }
+  }
+  return OkStatus();
+}
+
+bool ZenLayoutRewritePass::ZenOpRewritePass(std::unique_ptr<Graph> *g) {
+  bool result = false;
+  CHECK_NOTNULL(g);
+
+  // Before we proceed further for Zen Op rewrites first the graph shall be
+  // checked for inference ops only as TF-ZenDNN currently does not support
+  // training, it supports inference only.
+  TF_CHECK_OK(AreAllInferenceOps(g));
+
+  std::vector<Node *> order;
+
+  DumpGraph("\nBefore ZenRewritePass:\n", &**g);
+
+  // Two passes of Graph optimization:
+
+  // First pass implements Basic Fusion Eg. Conv2D-Bias-Relu.
+  result = ZenOpUpdate(g);
+  if (!result) {
+    VLOG(1) << "ZenLayoutRewritePass::ZenOpRewritePass: No opportunity for Zen "
+            << "op conversion found";
+  }
+  // Second Pass - Enable Fused Optimizations
+  // Enable Advanced Graph Optimizations
+  GetReversePostOrder(**g, &order);
+  for (Node *n : order) {
+    if (!n->IsOp() || !CanOpRunOnCPUDevice(n)) {
+      continue;
+    }
+    // Fused Optimizations
+
+    // Check and perform Pad fusion with FusedConv2D/Conv2D (Removes Pad op and
+    // expects n to be Pad op).
+    if (ZenFusePadConv(g, n, "_ZenFusedConv2D", "Pad")) {
+      VLOG(1) << "ZenLayoutRewritePass::ZenOpRewritePass: "
+              << "FusedConvPad Successful";
+    } else if (ZenFusePadConv(g, n, "_ZenConv2D", "Pad")) {
+      VLOG(1) << "ZenLayoutRewritePass::ZenOpRewritePass: ConvPad Successful";
+    }
+  }
+  // Update Zen op.
+  result = ZenOpUpdate(g);
+  if (!result) {
+    VLOG(1) << "ZenLayoutRewritePass::ZenOpRewritePass: No instance of "
+            << "FuseBatchNorm found.";
```

Comment:
Why is this error message only about FusedBatchNorm?

Reply:
We have re-phrased the comments and log for the graph optimization passes. Thank you!